24_clustering.Rmd

```diff
@@ -17,9 +18,127 @@ features/samples that share pattern.  For example, with clustering, we
 
+## Introduction to bluster package
+
+*bluster* is a Bioconductor package specialized in clustering. It offers 
+multiple algorithms such as hierarchical clustering, DBSCAN, K-means, 
+amongst others. The first thing to do when using this package is to load 
+it. It's also possible to use `transformCounts` to transform the data 
+for the clustering.
+```{r load-bluster}
+# Load dependencies
+if (!require(bluster)) {
+  install.packages("bluster")
+  library(bluster)
+}
+
+# Optional: Apply transformation
+tse <- transformCounts(tse, method = "relabundance")
+```
+
+The clustering can be done on the features or the samples.
+
+### Samples clustering
+
+To cluster the samples, we simply need to transpose the assay so that 
+the samples are in rows.
+```{r sample-clust-prep}
+assay <- t(assay(tse, "relabundance"))
+```
+
+Then, to do the clustering, use the `clusterRows` function. Depending 
+on its parameters, it will perform a different algorithm. Here, we'll 
+perform a hierarchical clustering.
+```{r bluster-sample-hclust1}
+# Simple use of the hierarchical clustering which sets the cut height to
+# half the dendrogram height.
+hclust.out <- clusterRows(assay, HclustParam())
+
+# Add data to the TSE
+colData(tse)$clusters <- hclust.out
+
+# Checking the result
+hclust.out
+```
+
+Once the clustering on the samples is done, we can also plot the clusters.
+```{r bluster-sample-plot}
+library(scater)
+# Add the MDS dimensions for plotting
+tse <- runMDS(tse,
+  assay.type = "relabundance", FUN = vegan::vegdist,
+  method = "bray"
+)
+
+# Plot the clusters
+plotReducedDim(tse, "MDS", colour_by = "clusters")
+```
+
+To change the clustering parameters, simply change the parameters in 
+`HclustParam` function. To see the different parameters, check the 
+[HclustParam documentation](https://rdrr.io/github/LTLA/bluster/man/HclustParam-class.html).
+```{r bluster-sample-hclust2}
+# More complex use of the algorithm with another method and cut height
+hclust.out <- clusterRows(assay, HclustParam(method = "complete", cut.dynamic = TRUE))
+hclust.out
+```
+
+Sometimes we need more information from the clustering, for example to 
+plot the dendogram. In that case we use the `full` parameter.
+```{r bluster-sample-dendo}
+# Get full hierarchical clustering information
+hclust.out <- clusterRows(assay, HclustParam(), full = TRUE)
+
+# Add data to the TSE
+colData(tse)$clusters <- hclust.out$clusters
+
+# Get the dendrogram object
+dendro <- as.dendrogram(hclust.out$objects$hclust)
+
+# Plot the dendrogram
+plot(dendro)
+```
+
+Alternatively, it's also possible to use the `clusterCells` function from 
+the *scran* package.This function uses directly the tree as a parameter 
+instead of the assay itself. Here, we will use the PAM algorithm.
+```{r scran-sample-clust}
+if (!require(scran)) {
+  BiocManager::install("scran")
+  library(scran)
+}
+pam.out <- clusterCells(tse,
+  assay.type = "relabundance",
+  BLUSPARAM = PamParam(centers = 5)
+)
+```
```

Comment:
I think this kind of wrapper could be nice

Reply:
The name for our purposes could be clusterSamples.

If DMM would not go to _bluster_ (we can check later), then it could be integrated in this kind of mia wrapper.